Add a Dynamo deepdive to documentation

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122305

This supersedes the previous `Guards Overview" as a more comprehensive
approach to most of the main topics within Dynamo.

In the future, we could add specific sections for each of the topics
discussed here.

lezcano committed:
Add a Dynamo deepdive to documentation

This supersedes the previous `Guards Overview" as a more comprehensive
approach to most of the main topics within Dynamo.

In the future, we could add specific sections for each of the topics
discussed here.

[ghstack-poisoned]

lezcano committed:
Update on "Add a Dynamo deepdive to documentation"


This supersedes the previous `Guards Overview" as a more comprehensive
approach to most of the main topics within Dynamo.

In the future, we could add specific sections for each of the topics
discussed here.

[ghstack-poisoned]

lezcano committed:
Update on "Add a Dynamo deepdive to documentation"


This supersedes the previous `Guards Overview" as a more comprehensive
approach to most of the main topics within Dynamo.

In the future, we could add specific sections for each of the topics
discussed here.

[ghstack-poisoned]

msaroufim commented on docs/source/torch.compiler_dynamo_deepdive.rst:
```diff
@@ -0,0 +1,873 @@
+
+All the inputs and intermediary elements of the FX graph are
+``fx.Node``\ s. In Dynamo, ``fx.Node``\ s are wrapped in
+``fx.Proxy``\ s, which record operations done on them in the graph. We
```
nit: can you elaborate what record operations done on them means?

msaroufim commented on docs/source/torch.compiler_dynamo_deepdive.rst:
```diff
@@ -0,0 +1,873 @@
+Making Dynamo Sound: Guards
+---------------------------
+
+At this point, we have a way to trace linear programs. And for that, we
```
what's a linear program?

msaroufim commented on docs/source/torch.compiler_dynamo_deepdive.rst:
```diff
@@ -0,0 +1,873 @@
+
+.. code:: python!
+
+   ___check_type_id(L['b'], 94334122025024)
```
where does the 9433.. number come from?

msaroufim commented on docs/source/torch.compiler_dynamo_deepdive.rst:
```diff
@@ -0,0 +1,873 @@
+
+   ___check_type_id(L['l'], 94439025877664)
+   len(L['l']) == 2
+   hasattr(L['x'], '_dynamo_dynamic_indices') == False
```
worth elaborating in a few more lines what this is

lezcano replied:
Just removed that line to avoid distracting the reader.

msaroufim commented on docs/source/torch.compiler_dynamo_deepdive.rst:
```diff
@@ -0,0 +1,873 @@
+   mul_1 (concrete): (8, 3)
+
+where we can see that the first dimension of the two tensor args is
+dynamic.
```
dynamic because there's symbolic values s0

msaroufim reviewed (approved):
really enjoyed reading this, thank you for writing it

lezcano committed:
Update on "Add a Dynamo deepdive to documentation"


This supersedes the previous `Guards Overview" as a more comprehensive
approach to most of the main topics within Dynamo.

In the future, we could add specific sections for each of the topics
discussed here.

[ghstack-poisoned]

lezcano committed:
Update on "Add a Dynamo deepdive to documentation"


This supersedes the previous `Guards Overview" as a more comprehensive
approach to most of the main topics within Dynamo.

In the future, we could add specific sections for each of the topics
discussed here.

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge

lezcano committed:
Update on "Add a Dynamo deepdive to documentation"


This supersedes the previous `Guards Overview" as a more comprehensive
approach to most of the main topics within Dynamo.

In the future, we could add specific sections for each of the topics
discussed here.

[ghstack-poisoned]
